# INTRODUCTION TO NunPy

In [1775]:
import numpy as np

print("NumPy Version:", np.__version__)

NumPy Version: 2.5.2


In [1776]:
# Using 1D Array - A single sequence of values (Identified by Number of square Bracket [] used.)
daily_revenue = np.array([45000, 62000, 38000, 71000, 55000, 49000, 83000, 14000])

print(daily_revenue)
print('Type:', type(daily_revenue))  # Used to tell us what kind of Python Object something is.
print("Shape:", daily_revenue.shape) # Used to check the number of columns & rows in a data.
print('Dtype:', daily_revenue.dtype) # Used to tell us the data type of the elements inside a NumPy array
print("Ndim:", daily_revenue.ndim)   # Used to check Number Dimensions of the data.

print()

#  Using 2D Array - Rows and Colunms values (Identified by Number of square Bracket [[]] used.)
# Weekly revenue by product category (Electronics, Fashion, Food, Drinks)
weekly_revenue = np.array([
    [850.45, 120.01, 420.09, 560.11],  # Week 1
    [150.81, 910.06, 380.03, 200.07],  # Week 2
    [780.78, 550.55, 90.56, 100.21],   # Week 3
    [180.09, 100.54, 610.07, 131.59],  # Week 4
    [200.45, 155.55, 785.25, 171.51],  # Week 5
])

print(weekly_revenue)
print('Shape:', weekly_revenue.shape)
# print("Type:", type(weekly_revenue))
print('Ndim:', weekly_revenue.ndim)
print("Dtype:", weekly_revenue.dtype)

[45000 62000 38000 71000 55000 49000 83000 14000]
Type: <class 'numpy.ndarray'>
Shape: (8,)
Dtype: int64
Ndim: 1

[[850.45 120.01 420.09 560.11]
 [150.81 910.06 380.03 200.07]
 [780.78 550.55  90.56 100.21]
 [180.09 100.54 610.07 131.59]
 [200.45 155.55 785.25 171.51]]
Shape: (5, 4)
Ndim: 2
Dtype: float64


In [1777]:
# To Load Dataset in NumPy
import csv

# Load our orders dataset manually (we'll use pd.read_csv())
orders = []
with open("orders.csv", encoding='utf-8-sig') as f:
    reader = csv.DictReader(f)
    for row in reader:
        orders.append(row)

print(f"Loaded {len(orders)} orders")
print("Columns:", list(orders[0].keys()))

Loaded 500 orders
Columns: ['order_id', 'customer_name', 'phone', 'email', 'state', 'city', 'product_category', 'product_name', 'quantity', 'unit_price', 'total_amount', 'order_date', 'delivery_status', 'payment_method']


In [1778]:
# Extract total_amount as a NumPy array
# Some Rows have missing total_amount - skip those for now
import numpy as np

total_amounts = np.array([
    float(row["total_amount"])
    for row in orders
    if row["total_amount"] != ""
])

print("Array Shape:", total_amounts.shape)
print("First 11 values:", total_amounts[:11])

Array Shape: (494,)
First 11 values: [ 10716.  35067.  23668.  47676.  80190. 104930.   1716.  96036.  13745.
 285000. 173730.]


In [1779]:
# Apply 7.5% VAT to every order — no loop needed
vat = total_amounts * 0.075
total_with_vat = total_amounts + vat

print("First 6 Original Amounts: ", total_amounts[:6])
print("First 6 VAT Amounts:      ", vat[:6].round(2))
print("First 6 Total with VAT:   ", total_with_vat[:6].round(2))

print()
print()

# Convert Naira to USD (1 USD = 1,600 NGN)
usd_amounts = total_amounts / 1600

print("First 6 Amounts in USD:", usd_amounts[:6].round(2))

First 6 Original Amounts:  [ 10716.  35067.  23668.  47676.  80190. 104930.]
First 6 VAT Amounts:       [ 803.7  2630.02 1775.1  3575.7  6014.25 7869.75]
First 6 Total with VAT:    [ 11519.7   37697.02  25443.1   51251.7   86204.25 112799.75]


First 6 Amounts in USD: [ 6.7  21.92 14.79 29.8  50.12 65.58]


#### INDEXING & SLICING IN NumPy

In [1780]:
# SINGLE ELEMENTS
print("First Order: ", total_amounts[0])
print("Last Order:  ", total_amounts[-1])
print("5th Order:   ", total_amounts[4])

print()
print()

# SLICES
print("First 7 Orders:  ", total_amounts[:7])
print("Orders 10 to 15: ", total_amounts[10:16])
print("Last 5 Orders:   ", total_amounts[-5:])

print()
print()

# 2D(2 DIMENSION) INDEXING ON THE weekly_revenue FRID FROM ABOVE
print("Week 2, Fashion Revenue:      ", weekly_revenue[1, 1])       # FROM Row 1, Column 1
print("All of Week 3:                ", weekly_revenue[2,:])        # FROM Entire Row 2
print("All Electronics (column 0):   ", weekly_revenue[:, 0])       # FROM Entire Column 0
print("Week 2 and 3, Fashion & Food: ", weekly_revenue[1:4, 1:3])   # FOM Row 1 and 2, & Column 1 and 2.


First Order:  10716.0
Last Order:   20939.0
5th Order:    80190.0


First 7 Orders:   [ 10716.  35067.  23668.  47676.  80190. 104930.   1716.]
Orders 10 to 15:  [173730.  21406.  59560.  86528.  41844.  13072.]
Last 5 Orders:    [ 3974. 23668. 32620.  8084. 20939.]


Week 2, Fashion Revenue:       910.06
All of Week 3:                 [780.78 550.55  90.56 100.21]
All Electronics (column 0):    [850.45 150.81 780.78 180.09 200.45]
Week 2 and 3, Fashion & Food:  [[910.06 380.03]
 [550.55  90.56]
 [100.54 610.07]]


#### STATISTICAL FUNCTIONS IN NumPy

In [1781]:
print("Total Revenue:     NGN", f"{np.sum(total_amounts):,.0f}")
print("Average Order:     NGN", F"{np.mean(total_amounts):,.0f}")
print("Median Order:      NGN", f"{np.median(total_amounts):,.0f}")
print("Smallest Order:    NGN", f"{np.min(total_amounts):,.0f}")  # min = Minimum
print("Largest Order:     NGN", f"{np.max(total_amounts):,.0f}")  # max = Maximum
print("5th Deviation:     NGN", f"{np.std(total_amounts):,.0f}")
print("75th Percentile:   NGN", f"{np.percentile(total_amounts, 75):,.0f}")

print()
print()

# Why is the mean much higher than the median?
# If MEAN is Higher than Median - It is Called Outliers
# Let's find those extreme orders
mean_val = np.mean(total_amounts)
median_val = np.median(total_amounts)

print(f"Mean:         NGN {mean_val:,.0f}")
print(f"Median:       NGN {median_val:,.0f}")
print(f"Differance:   NGN {mean_val - median_val:,.0f}")

# Orders Above NGN 200,000 - The Outliers Pulling The Mean up
outliers = total_amounts[total_amounts > 200000]
print("Extreme Orders Above NGN 200,000:", outliers)


Total Revenue:     NGN 21,748,459
Average Order:     NGN 44,025
Median Order:      NGN 27,182
Smallest Order:    NGN 842
Largest Order:     NGN 312,000
5th Deviation:     NGN 44,232
75th Percentile:   NGN 61,034


Mean:         NGN 44,025
Median:       NGN 27,182
Differance:   NGN 16,843
Extreme Orders Above NGN 200,000: [285000. 312000. 223600. 200875.]


#### THE AXIS PARAMETER


In [1782]:
# Use the weekly_revenue grid: shape (5 weeks, 3 categories)
# Categories: Electronics, Fashion, Food & Drinks
print("Grid Shape:", weekly_revenue.shape)
print()

# Total Per CATEGORY - Sum Down the Rows (axis=0)
category_total = np.sum(weekly_revenue, axis=0)
print("TOTAL PER CATEGORY (axis=0):")
print(f"   Electronics:  NGN {category_total[0]:,.2f}")
print(f"   Fashion:      NGN {category_total[1]:,.2f}")
print(f"   Food:         NGN {category_total[2]:,.2f}")
print(f"   Drinks:       NGN {category_total[3]:,.2f}")
print()

# Total per WEEK - Sum across the Columns (axis=1)
weekly_total = np.sum(weekly_revenue, axis=1)
print("TOTAL PER WEEK (axis=1):")
for i, total in enumerate(weekly_total, 1):
    print(f"   Week {i}: NGN {total:,.2f}")

print()

# Which Week had the Best Performance Per Category?
best_week_per_category = np.argmax(weekly_revenue, axis=0) + 1 # argmax is Argument of the maximum value in a function oe array. 
                                                                # +1 to make it Week 1-based.
print("Best Week For Electronics: Week", best_week_per_category[0])
print("Best Week For Fashion:     Week", best_week_per_category[1])
print("Best Week For Food:        Week", best_week_per_category[2])
print("Best Week For Drinks:      Week", best_week_per_category[3])

Grid Shape: (5, 4)

TOTAL PER CATEGORY (axis=0):
   Electronics:  NGN 2,162.58
   Fashion:      NGN 1,836.71
   Food:         NGN 2,286.00
   Drinks:       NGN 1,163.49

TOTAL PER WEEK (axis=1):
   Week 1: NGN 1,950.66
   Week 2: NGN 1,640.97
   Week 3: NGN 1,522.10
   Week 4: NGN 1,022.29
   Week 5: NGN 1,312.76

Best Week For Electronics: Week 1
Best Week For Fashion:     Week 2
Best Week For Food:        Week 5
Best Week For Drinks:      Week 1


#### BOOLEAN INDEXING (FILTERING)

In [1783]:
# Step 1: A Condition Produces a True/False Array
high_value_mask = total_amounts > 20000

print("Boolean Mask (first 10):", high_value_mask[:10])
print("Number of High-Value Orders:", np.sum(high_value_mask))  # True Counts as 1 

print()

high_value_mask = total_amounts > 25000

print("Boolean Mask (first 10):", high_value_mask[:10])
print("Number of High-Value Orders:", np.sum(high_value_mask))  # True Counts as 1 

print()
print()

# Step 2: Use the Mask to Filter
high_value_order = total_amounts[total_amounts > 20000]

print("HIGH-VALUE ORDERS (>NGN 20,000):")
print(high_value_order)
print()
print(f"COUNT:   {len(high_value_order)} Orders")
print(f"AVERAGE: NGN {np.mean(high_value_order):,.0f}")

print()
print()

high_value_order = total_amounts[total_amounts > 25000]

print("HIGH-VALUE ORDERS (>NGN 25,000):")
print(high_value_order)
print()
print(f"COUNT:   {len(high_value_order)} Orders")
print(f"AVERAGE: NGN {np.mean(high_value_order):,.2f}")

Boolean Mask (first 10): [False  True  True  True  True  True False  True False  True]
Number of High-Value Orders: 309

Boolean Mask (first 10): [False  True False  True  True  True False  True False  True]
Number of High-Value Orders: 261


HIGH-VALUE ORDERS (>NGN 20,000):
[ 35067.  23668.  47676.  80190. 104930.  96036. 285000. 173730.  21406.
  59560.  86528.  41844.  83098.  66216.  31803.  20265.  23736.  71652.
  23544.  23264.  38192.  46251.  54066.  35360.  34570.  98492. 136045.
  95045.  97488.  32370.  32620.  52986.  20745.  20560. 144835.  51065.
  43230.  22022.  21983. 312000.  49815.  43068.  24784.  60942. 132009.
  91122.  75284.  84890.  66420.  68625.  23265.  47170.  50583.  27316.
  76872.  97105.  40648.  31572. 114645.  38943.  23285.  41430. 175196.
  23298.  86220.  37776.  29877.  23247.  80918. 142608.  26164.  65298.
  64468.  22861.  25983.  61100.  44930.  68865.  69248. 137710.  53120.
  44537.  53208.  36070.  32664.  59790.  21932.  66232. 159736.  6

#### PUTTING IT ALL TOGETHER

**Business question:** *"What percentage of our orders are high-value (above NGN 20,000), and how much of total revenue do they represent?"*

In [1784]:
all_orders      = total_amounts
high_value      = total_amounts[total_amounts > 20000]
standard_value  = total_amounts[total_amounts <= 20000]

total_rev     = np.sum(all_orders)
high_rev      = np.sum(high_value)
standard_rev  = np.sum(standard_value)

print("====== ORDER VALUE ANALYSIS ======")
print(f"Total Orders:            {len(all_orders)}")
print(f"High-Value Orders:       {len(high_value)} ({len(high_value)/len(all_orders)*100:.2f}%)")
print(f"Standard Orders:         {len(standard_value)} ({len(standard_value)/len(all_orders)*100:.2f}%)")
print()
print(f"Total Revenue:           NGN {total_rev:,.0f}")
print(f"High-Value Revenue:      NGN {high_rev:,.0f} ({high_rev/total_rev*100:.2f}% of total)")
print(f"Standard Revenue:        NGN {standard_rev:,.0f} ({standard_rev/total_rev*100:.2f}% of total)")

====== ORDER VALUE ANALYSIS ======
Total Orders:            494
High-Value Orders:       309 (62.55%)
Standard Orders:         185 (37.45%)

Total Revenue:           NGN 21,748,459
High-Value Revenue:      NGN 19,795,355 (91.02% of total)
Standard Revenue:        NGN 1,953,104 (8.98% of total)


# INTRODUCTION TO Pandas

In [1785]:
import pandas as pd

print("Pandas Version:", pd.__version__)

Pandas Version: 3.0.5


In [1786]:
# A Series - A Single Labelled Column
revenue_by_state = pd.Series(
    [842000, 531000, 394000, 287000, 115000, 215000],
    index=["Lagos", "Abuja", "Rivers", "Kano", "Osun", "Oyo"]
)
print(revenue_by_state)
print()
print("Lagos Revenue: NGN", revenue_by_state["Lagos"])
print("Dtype:", revenue_by_state.dtype)
print("Index:", revenue_by_state.index.tolist())

Lagos     842000
Abuja     531000
Rivers    394000
Kano      287000
Osun      115000
Oyo       215000
dtype: int64

Lagos Revenue: NGN 842000
Dtype: int64
Index: ['Lagos', 'Abuja', 'Rivers', 'Kano', 'Osun', 'Oyo']


In [1787]:
# A DataFrame — built from a dictionary (keys = column names)
sample_orders = pd.DataFrame({
    "order_id":    ["ORD-001", "ORD-002", "ORD-003", "ORD-004", "ORD-005"],
    "customer":    ["Ngozi Okafor", "Emeka Bello", "Fatima Musa", "Tunde Lawal", "Aisha Garba"],
    "state":       ["Lagos", "Abuja", "Kano", "Rivers", "Kaduna"],
    "category":    ["Electronics", "Fashion", "Food & Drinks", "Electronics", "Beauty & Health"],
    "amount":      [4500, 12000, 800, 33000, 5600],
    "delivered":   [True, True, False, True, False],
})

sample_orders

,order_id,customer,state,category,amount,delivered
0,ORD-001,Ngozi Okafor,Lagos,Electronics,4500,True
1,ORD-002,Emeka Bello,Abuja,Fashion,12000,True
2,ORD-003,Fatima Musa,Kano,Food & Drinks,800,False
3,ORD-004,Tunde Lawal,Rivers,Electronics,33000,True
4,ORD-005,Aisha Garba,Kaduna,Beauty & Health,5600,False


##### KEY DATAFRAME ATTRIBUTES IN PANDAS

In [1788]:
# Key DataFrame Attributes
print("Shapes: ", sample_orders.shape)
print("Columns:", sample_orders.columns.tolist())
print("Dtypes:\n", sample_orders.dtypes)
print()

# Each Column in a Series
print("Type of a Column:", type(sample_orders["amount"]))

Shapes:  (5, 6)
Columns: ['order_id', 'customer', 'state', 'category', 'amount', 'delivered']
Dtypes:
 order_id       str
customer       str
state          str
category       str
amount       int64
delivered     bool
dtype: object

Type of a Column: <class 'pandas.Series'>


In [1789]:
# To Load Dataset in Pandas
# Load the full orders dataset
df = pd.read_csv("orders.csv", encoding="utf-8-sig")

print(f"Loaded: {df.shape[0]} rows x {df.shape[1]} columns")

Loaded: 500 rows x 14 columns


##### KEY DATAFRAME METHODS IN PANDAS & UNDERSTANDING THE STRUCTURE

In [1790]:
# Always Start With head() - First 5 Rows
df. head()

,order_id,customer_name,phone,email,state,city,product_category,product_name,quantity,unit_price,total_amount,order_date,delivery_status,payment_method
0,ORD-0001,Rukayat Obi,7.086001e+09,rukayatobi@hotmail.com,Abuja,Maitama,Fashion,Handbag,2,₦5358,10716.0,2024-02-08,Shipped,Cash on Delivery
1,ORD-0002,Emeka Obiora,8.066156e+09,emekaobiora@hotmail.com,Oyo,Ogbomosho,Home & Kitchen,Laundry Basket,3,₦11689,35067.0,2024-02-03,Processing,USSD
2,ORD-0003,Adaeze Abubakar,8.131648e+09,adaeze_abubakar27@gmail.com,OGUN,Ota,Fashion,Women's Dress,1,₦23668,23668.0,2024-02-20,Cancelled,Bank Transfer
3,ORD-0004,Rukayat Obiora,8.130564e+09,rukayatobiora@hotmail.com,anambra,Awka,Home & Kitchen,Curtains Set,3,₦15892,47676.0,2024-01-13,Cancelled,Opay Wallet
4,ORD-0005,Segun Chukwu,NaN,segunchukwu21@gmail.com,KANO,Fagge,Fashion,Sunglasses,5,₦16038,80190.0,2024-06-13,Shipped,Cash on Delivery


In [1791]:
# tail() - Last 5 Rows
df.tail()

,order_id,customer_name,phone,email,state,city,product_category,product_name,quantity,unit_price,total_amount,order_date,delivery_status,payment_method
495,ORD-0496,Ngozi Eze,8.065915e+09,ngozi.eze@yahoo.com,Enugu,Oji River,Beauty & Health,Lip Gloss Set,1,₦3974,3974.0,2024-11-17,Delivered,Cash on Delivery
496,ORD-0497,Adaeze Abubakar,8.131648e+09,adaeze_abubakar27@gmail.com,OGUN,Ota,Fashion,Women's Dress,1,₦23668,23668.0,2024-02-20,Cancelled,Bank Transfer
497,ORD-0498,Chidi Eze,8.160748e+09,chidi_eze76@hotmail.com,ogun,Ilaro,Beauty & Health,Face Moisturiser,2,₦16310,32620.0,2024-08-23,Delivered,Opay Wallet
498,ORD-0499,Oluwaseun Nwosu,7.065652e+09,oluwaseun.nwosu71@gmail.com,Kano,Tarauni,Home & Kitchen,Dish Rack,1,₦8084,8084.0,2024-10-28,Cancelled,USSD
499,ORD-0500,Musa Nwosu,8.036357e+09,musa_nwosu84@outlook.com,Ogun,Ijebu-Ode,Fashion,Men's Shirt,1,₦20939,20939.0,2024-09-28,Processing,Opay Wallet


In [1792]:
# info() - Column Names, Dtypes, and Missing Values Counts
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   order_id          500 non-null    str    
 1   customer_name     500 non-null    str    
 2   phone             492 non-null    float64
 3   email             500 non-null    str    
 4   state             500 non-null    str    
 5   city              500 non-null    str    
 6   product_category  500 non-null    str    
 7   product_name      500 non-null    str    
 8   quantity          500 non-null    int64  
 9   unit_price        500 non-null    str    
 10  total_amount      494 non-null    float64
 11  order_date        500 non-null    str    
 12  delivery_status   500 non-null    str    
 13  payment_method    500 non-null    str    
dtypes: float64(2), int64(1), str(11)
memory usage: 54.8 KB


In [1793]:
# describe() - Statistical Summary of Numeric Columns
df.describe()

,phone,quantity,total_amount
count,4.920000e+02,500.000000,494.000000
mean,7.962036e+09,2.992000,44025.220648
std,7.204476e+08,1.412773,44276.571833
min,7.030167e+09,1.000000,842.000000
25%,7.083678e+09,2.000000,13469.000000
50%,8.069125e+09,3.000000,27182.000000
75%,8.164321e+09,4.000000,61034.250000
max,9.059320e+09,5.000000,312000.000000


In [1794]:
# isnull() - Used to Check Missing Values (Each Column Has) in a DataFrame. It Returns TRUE where a value is Missing & FALSE where a value is Present.
# isnull() - Checks/Identifies Missing Values(True or False).
df.isnull()

,order_id,customer_name,phone,email,state,city,product_category,product_name,quantity,unit_price,total_amount,order_date,delivery_status,payment_method
0,False,False,False,False,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False,False,False,False,False
4,False,False,True,False,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
495,False,False,False,False,False,False,False,False,False,False,False,False,False,False
496,False,False,False,False,False,False,False,False,False,False,False,False,False,False
497,False,False,False,False,False,False,False,False,False,False,False,False,False,False
498,False,False,False,False,False,False,False,False,False,False,False,False,False,False


In [1795]:
# isnull().sum() - Counts the Missing Values.
df.isnull().sum()

order_id            0
customer_name       0
phone               8
email               0
state               0
city                0
product_category    0
product_name        0
quantity            0
unit_price          0
total_amount        6
order_date          0
delivery_status     0
payment_method      0
dtype: int64

In [1796]:
# What we noticed from info():
# 1. phone has 492 non-null (8 missing)
# 2. total_amoount has 494 non-null (6 missing)
# 3. unit_price is 'object' - It should be a number (has ₦ prefix)
# 4. order_date is 'object' - It should be a datetime

# Let's confirm the unit_price issue
print("Unit_Price Samples:")
print(df["unit_price"].head(10).tolist())

print()
print()

# Key Observations from describe():
mean_val   = df["total_amount"].mean()
median_val = df["total_amount"].median()
max_val    = df["total_amount"].max()

print(f"Mean Order Value:     NGN {mean_val:,.0f}")
print(f"Median Order Value:   NGN {median_val:,.0f}")
print(f"Max Order Value:      NGN {max_val:,.0f}")
print()
print(f"Gap (mean - median):  NGN {mean_val - median_val:,.0f}")
print("→ Outliers are pulling the mean up")

Unit_Price Samples:
['₦5358', '₦11689', '₦23668', '₦15892', '₦16038', '₦20986', '₦1716', '₦24009', '₦2749', '₦10661']


Mean Order Value:     NGN 44,025
Median Order Value:   NGN 27,182
Max Order Value:      NGN 312,000

Gap (mean - median):  NGN 16,843
→ Outliers are pulling the mean up


##### SELECTING COLUMNS

In [1797]:
# To Select Single Column ⇀ Series
States = df["state"]
print("Type:", type(States))
print(States.head(8))

print()
print()

# To Select Multiple Column ⇀ DataFrame (Double Brackets)
Subset = df[["customer_name", "state", "product_category", "total_amount"]]
print("Type:", type((Subset)))
Subset.head(8)

print()
print()

# Useful Series Operations on a Column
print("Unique Categories:")
print(df["product_category"].unique())
print()

print("Orders Per Category:")
print(df["product_category"].value_counts())
print()

print("Orders Per Delivery Status:")
print(df["delivery_status"].value_counts())

Type: <class 'pandas.Series'>
0      Abuja
1        Oyo
2       OGUN
3    anambra
4       KANO
5      delta
6    Anambra
7      enugu
Name: state, dtype: str


Type: <class 'pandas.DataFrame'>


Unique Categories:
<StringArray>
['Fashion', 'Home & Kitchen', 'Electronics', 'Food & Drinks',
 'Beauty & Health']
Length: 5, dtype: str

Orders Per Category:
product_category
Fashion            111
Electronics        104
Food & Drinks       99
Beauty & Health     95
Home & Kitchen      91
Name: count, dtype: int64

Orders Per Delivery Status:
delivery_status
Delivered     246
Cancelled      88
Shipped        83
Processing     83
Name: count, dtype: int64


##### FILTERING ROWS

In [1798]:
# For All Orders From Abuja
# Note: State Column has Inconsistent Casing 
abuja_orders = df[df["state"] == "Abuja"]
print(f"Orders With State == 'Abuja' {len(abuja_orders)}")

# Why might this be fewer than expected?
print("\nAll Unique State Values (Showing the Inconsistency):")
print(sorted(df["state"].unique()))

Orders With State == 'Abuja' 31

All Unique State Values (Showing the Inconsistency):
['ABUJA', 'ANAMBRA', 'Abuja', 'Anambra', 'DELTA', 'Delta', 'ENUGU', 'Enugu', 'KADUNA', 'KANO', 'Kaduna', 'Kano', 'LAGOS', 'Lagos', 'OGUN', 'OYO', 'Ogun', 'Oyo', 'RIVERS', 'Rivers', 'abuja', 'anambra', 'delta', 'enugu', 'kaduna', 'kano', 'lagos', 'ogun', 'oyo', 'rivers']


In [1799]:
# High-Value Orders - Above NGN 30,000
high_value = df[df["total_amount"] > 30000]
print(f"High-Value Orders (>30000): {len(high_value)}") 
high_value[["order_id", "customer_name", "product_category", "total_amount"]].head(10)

High-Value Orders (>30000): 233


,order_id,customer_name,product_category,total_amount
1,ORD-0002,Emeka Obiora,Home & Kitchen,35067.0
3,ORD-0004,Rukayat Obiora,Home & Kitchen,47676.0
4,ORD-0005,Segun Chukwu,Fashion,80190.0
5,ORD-0006,Ifeanyi Obiora,Electronics,104930.0
7,ORD-0008,Chidi Salami,Fashion,96036.0
10,ORD-0011,Tunde Adeyemi,Home & Kitchen,285000.0
11,ORD-0012,Musa Usman,Electronics,173730.0
13,ORD-0014,Halima Usman,Fashion,59560.0
14,ORD-0015,Ngozi Okafor,Electronics,86528.0
15,ORD-0016,Ngozi Salami,Fashion,41844.0


In [1800]:
# Multiple Conditions - Electronics Orders Above NGN 10,000
electronics_high = df[
    (df["product_category"] == "Electronics") &
    (df["total_amount"] > 10000)
]
print(f"electronics Orders > 10000: {len(electronics_high)}")
electronics_high[["order_id", "state", "product_name", "total_amount"]].head(8)

electronics Orders > 10000: 98


,order_id,state,product_name,total_amount
5,ORD-0006,delta,USB Charger,104930.0
11,ORD-0012,Lagos,Bluetooth Speaker,173730.0
14,ORD-0015,Oyo,USB Charger,86528.0
17,ORD-0018,ENUGU,Laptop Stand,83098.0
26,ORD-0027,lagos,Wireless Earbuds,23544.0
27,ORD-0028,Enugu,Phone Case,23264.0
29,ORD-0030,Ogun,Webcam,46251.0
30,ORD-0031,Delta,USB Charger,54066.0


In [1801]:
# .isin() - Match Any Value in a List
# Deliverec or Shipped Orders
active_orders = df[df["delivery_status"].isin(["Delivered", "Shipped"])]
print(f"Active Orders (Delivered or Shipped): {len(active_orders)}")
print(active_orders["delivery_status"].value_counts())

Active Orders (Delivered or Shipped): 329
delivery_status
Delivered    246
Shipped       83
Name: count, dtype: int64


##### LOC AND ILOC

In [1802]:
# iloc - Used based on Integer Position (Like NumPy)
print("  ====== HOW TO USE ILOC METHOD ======")
print("First Row (iloc[0]):")
print(df.iloc[0])
print()

print("First 3 Rows, First 4 Columns (iloc[0:3], [0:4]):")
print(df.iloc[0:3, 0:4])

print()
print()

# loc - Used on Label-based Selection & Integer Position
print("  ====== HOW TO USE LOC METHOD ======")
print("Row 0, customer_name Column:")
print(df.loc[0, "customer_name"])
print()

print("Rows 0-4, order_id to state:")
print(df.loc[0:4, "order_id":"state"])



  ====== HOW TO USE ILOC METHOD ======
First Row (iloc[0]):
order_id                          ORD-0001
customer_name                  Rukayat Obi
phone                         7086001338.0
email               rukayatobi@hotmail.com
state                                Abuja
city                               Maitama
product_category                   Fashion
product_name                       Handbag
quantity                                 2
unit_price                           ₦5358
total_amount                       10716.0
order_date                      2024-02-08
delivery_status                    Shipped
payment_method            Cash on Delivery
Name: 0, dtype: object

First 3 Rows, First 4 Columns (iloc[0:3], [0:4]):
   order_id    customer_name         phone                        email
0  ORD-0001      Rukayat Obi  7.086001e+09       rukayatobi@hotmail.com
1  ORD-0002     Emeka Obiora  8.066156e+09      emekaobiora@hotmail.com
2  ORD-0003  Adaeze Abubakar  8.131648e+09  adae

In [1803]:
# loc with a condition - filter rows & select specific columns in one step
high_value_customers = df.loc[
    df["total_amount"] > 30000,
    ["customer_name", "state", "product_category", "total_amount"]
]

print(f"High-Value Customers ({len(high_value_customers)} orders):")
high_value_customers.head(10)

High-Value Customers (233 orders):


,customer_name,state,product_category,total_amount
1,Emeka Obiora,Oyo,Home & Kitchen,35067.0
3,Rukayat Obiora,anambra,Home & Kitchen,47676.0
4,Segun Chukwu,KANO,Fashion,80190.0
5,Ifeanyi Obiora,delta,Electronics,104930.0
7,Chidi Salami,enugu,Fashion,96036.0
10,Tunde Adeyemi,ABUJA,Home & Kitchen,285000.0
11,Musa Usman,Lagos,Electronics,173730.0
13,Halima Usman,RIVERS,Fashion,59560.0
14,Ngozi Okafor,Oyo,Electronics,86528.0
15,Ngozi Salami,anambra,Fashion,41844.0


##### SAVING RESULTS IN PANDAS

In [1804]:
high_value_customers.to_csv(
    "Data/High_Value_Orders.csv",
    index=False,
    encoding="utf-8-sig"
)

# Verify it Saved Correctly by Reloading
verify = pd.read_csv("Data/High_Value_Orders.csv", encoding="utf-8-sig")
print(f"Saved & Reloaded: {verify.shape[0]} rows x {verify.shape[1]} columns")
verify.head(6)

Saved & Reloaded: 233 rows x 4 columns


,customer_name,state,product_category,total_amount
0,Emeka Obiora,Oyo,Home & Kitchen,35067.0
1,Rukayat Obiora,anambra,Home & Kitchen,47676.0
2,Segun Chukwu,KANO,Fashion,80190.0
3,Ifeanyi Obiora,delta,Electronics,104930.0
4,Chidi Salami,enugu,Fashion,96036.0
5,Tunde Adeyemi,ABUJA,Home & Kitchen,285000.0


##### PUTTING IT ALL TOGETHER

**Business Question:** *"Give me a summary of Electronics orders - how many, total revenue, and average order value - broken down by payments method."*

##### WHAT DOES .copy() DO IN PANDAS
**.copy** is a Pandas DataFrame/Series Method.

**.copy** is Used to make a Seperate Copy of a DataFrame or Series.

The Main reason you use it is to make Changes to the New DataFrame without accidentally changing the Original One.

  For Example:

     high_value_orders = df[df["total_amount] > 250000].copy()

  Here:
  
      df ⇀ your original DataFrame
      df[df["total_amount] > 25000] ⇀ selects only orders above 25,000
      .copy() ⇀ creates an independent copy of those selected rows

KEY: **.copy** is like Photocopying a Document.


In [1805]:
# Step 1: Filter to Electronics only
electronics = df[df["product_category"] == "Electronics"].copy()
print(f"Electronics Orders: {len(electronics)}")

# Step 2: Drop Rows with missing total_amount
electonics = electronics.dropna(subset=["total_amount"])
print(f"After Dropping Missing Amounts: {len(electronics)}")

print()

# Step 3: Summary By Payment Method
payment_summary = electronics.groupby("payment_method")["total_amount"].agg(
    order_count="count",
    total_revenue="sum",
    avg_order="mean"
).round(0)

payment_summary

Electronics Orders: 104
After Dropping Missing Amounts: 104



,order_count,total_revenue,avg_order
payment_method,,,
Bank Transfer,22,1388965.0,63135.0
Card,21,2137990.0,101809.0
Cash on Delivery,23,1302280.0,56621.0
Opay Wallet,21,1949062.0,92812.0
USSD,16,976484.0,61030.0


##### UNDERSTANDING THE CODE BELOW:
**FROM CODE -- for method, row in payment_summary.itterrows():**

**What does .itterrows() mean?**

**.iterrows() Means:**

       Go through the DataFrame one Row at a Time. 
       .iterrows() takes them one row at a time.

**What is method?**

**method** 

       Gives the Row's Index (Your Payment Method)
**row**

      Gives all the Information in that Row.

      **KEY: **
      So for Bank Transfer:
          method is: Bank Transfer
      And:
          row contains:
          order_count    22
          total_revenue  1388965
          avg_order      63135
      That's why youcan do:
           row["total_revenue"] to get revenue.

**For the < & > Signs**

         **From Code -- {method:<20}** The < here does Not mean "Less than"
         It's part of Python's Formatting Syntax.

**<20 Means:**
          Put the Text on the Left and Give it a Space of 20 Characters.

      **FOR EXAMPLE:**
             f"{'card':<20}" -- Would make space for 20 Characters and put Card on the Left.
    
**>3 Means:** -- {int(row['order_count']):>3}

         Give me a Space of 3 Characters and put the Number on the right.
         This Helps the Numbers Line Up.

**>10,.0f** -- {row['total_revenue']:>10,.0f}

         There are actually three Formatting Instructions Here:
            - >10
                Right-Align the Number in a Space of 10 Characters. 
            - ,
                Add Commas to Large Numbers.
            - .0f
                Shows Zero Decimal PLaces.

**KEY:**

     Any where you see < any number & >any number Inside an f-String:
          - < any number Means
                  LEFT-ALIGN, THE NUMBER OF SPACES SPECIFIED.
          - > any number Means
                  RIGHT-ALIGN, THE NUMBER OF SPACES SPECIFIED.
        

In [1806]:
# Step 4: Format & Present
print("====== ELECTRONICS ORDERS BY PAYMENT METHOD ======")
for method, row in payment_summary.iterrows():
    print(f"  {method:<20} | Orders: {int(row['order_count']):>3} |"
          f"Revenue: NGN {row['total_revenue']:>10,.0f} | "
          f"Avg: NGN {row['avg_order']:>8,.0f}")

====== ELECTRONICS ORDERS BY PAYMENT METHOD ======
  Bank Transfer        | Orders:  22 |Revenue: NGN  1,388,965 | Avg: NGN   63,135
  Card                 | Orders:  21 |Revenue: NGN  2,137,990 | Avg: NGN  101,809
  Cash on Delivery     | Orders:  23 |Revenue: NGN  1,302,280 | Avg: NGN   56,621
  Opay Wallet          | Orders:  21 |Revenue: NGN  1,949,062 | Avg: NGN   92,812
  USSD                 | Orders:  16 |Revenue: NGN    976,484 | Avg: NGN   61,030


# CLEANING DATA


In [1807]:
# LOAD AND INSPECT DATA
import pandas as pd
import numpy as np

# LOAD RAW DATA
df = pd.read_csv("orders.csv", encoding='utf-8-sig')

print(f"Raw Dataset: {df.shape[0]} rows x {df.shape[1]} columns")
df.head(6)

Raw Dataset: 500 rows x 14 columns


,order_id,customer_name,phone,email,state,city,product_category,product_name,quantity,unit_price,total_amount,order_date,delivery_status,payment_method
0,ORD-0001,Rukayat Obi,7.086001e+09,rukayatobi@hotmail.com,Abuja,Maitama,Fashion,Handbag,2,₦5358,10716.0,2024-02-08,Shipped,Cash on Delivery
1,ORD-0002,Emeka Obiora,8.066156e+09,emekaobiora@hotmail.com,Oyo,Ogbomosho,Home & Kitchen,Laundry Basket,3,₦11689,35067.0,2024-02-03,Processing,USSD
2,ORD-0003,Adaeze Abubakar,8.131648e+09,adaeze_abubakar27@gmail.com,OGUN,Ota,Fashion,Women's Dress,1,₦23668,23668.0,2024-02-20,Cancelled,Bank Transfer
3,ORD-0004,Rukayat Obiora,8.130564e+09,rukayatobiora@hotmail.com,anambra,Awka,Home & Kitchen,Curtains Set,3,₦15892,47676.0,2024-01-13,Cancelled,Opay Wallet
4,ORD-0005,Segun Chukwu,NaN,segunchukwu21@gmail.com,KANO,Fagge,Fashion,Sunglasses,5,₦16038,80190.0,2024-06-13,Shipped,Cash on Delivery
5,ORD-0006,Ifeanyi Obiora,8.166270e+09,ifeanyiobiora14@yahoo.com,delta,Agbor,Electronics,USB Charger,5,₦20986,104930.0,2024-02-05,Delivered,Bank Transfer


In [1808]:
# Full Structural Inspection
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   order_id          500 non-null    str    
 1   customer_name     500 non-null    str    
 2   phone             492 non-null    float64
 3   email             500 non-null    str    
 4   state             500 non-null    str    
 5   city              500 non-null    str    
 6   product_category  500 non-null    str    
 7   product_name      500 non-null    str    
 8   quantity          500 non-null    int64  
 9   unit_price        500 non-null    str    
 10  total_amount      494 non-null    float64
 11  order_date        500 non-null    str    
 12  delivery_status   500 non-null    str    
 13  payment_method    500 non-null    str    
dtypes: float64(2), int64(1), str(11)
memory usage: 54.8 KB


In [1809]:
df.describe()

,phone,quantity,total_amount
count,4.920000e+02,500.000000,494.000000
mean,7.962036e+09,2.992000,44025.220648
std,7.204476e+08,1.412773,44276.571833
min,7.030167e+09,1.000000,842.000000
25%,7.083678e+09,2.000000,13469.000000
50%,8.069125e+09,3.000000,27182.000000
75%,8.164321e+09,4.000000,61034.250000
max,9.059320e+09,5.000000,312000.000000


In [1810]:
# full missing values inspection
df.isnull().sum()

order_id            0
customer_name       0
phone               8
email               0
state               0
city                0
product_category    0
product_name        0
quantity            0
unit_price          0
total_amount        6
order_date          0
delivery_status     0
payment_method      0
dtype: int64

In [1811]:
# Missing Value Counts & Percentages
missing_counts = df.isnull().sum()
missing_percent = (missing_counts / len(df) * 100).round(2)

missing_report = pd.DataFrame({
    "missing_count": missing_counts,
    "missing_percentage": missing_percent
})

print("Columns with Missing Values:")
missing_report[missing_report["missing_count"] > 0]

Columns with Missing Values:


,missing_count,missing_percentage
phone,8,1.6
total_amount,6,1.2


In [1812]:
# To Check Duplicate 
print(f"Duplicate Rows: {df.duplicated().sum()}")
print()

print("The Duplicated Rows:")
df[df.duplicated(keep=False)].sort_values("order_id").head(10)

Duplicate Rows: 0

The Duplicated Rows:


,order_id,customer_name,phone,email,state,city,product_category,product_name,quantity,unit_price,total_amount,order_date,delivery_status,payment_method


In [1813]:
# Type Issues - unit_price should be Numeric
print("unit_price dtype:", df["unit_price"].dtype)
print("unit_price_samples:", df["unit_price"].head(8).tolist())
print()

# Casting Issue - state column
print("Unique State Values:")
print(sorted(df["state"].unique()))

unit_price dtype: str
unit_price_samples: ['₦5358', '₦11689', '₦23668', '₦15892', '₦16038', '₦20986', '₦1716', '₦24009']

Unique State Values:
['ABUJA', 'ANAMBRA', 'Abuja', 'Anambra', 'DELTA', 'Delta', 'ENUGU', 'Enugu', 'KADUNA', 'KANO', 'Kaduna', 'Kano', 'LAGOS', 'Lagos', 'OGUN', 'OYO', 'Ogun', 'Oyo', 'RIVERS', 'Rivers', 'abuja', 'anambra', 'delta', 'enugu', 'kaduna', 'kano', 'lagos', 'ogun', 'oyo', 'rivers']


In [1814]:
# TO MAKE A COPY OF THE DATA
# Always work on a copy - Never modify the original DataFrame
df_clean = df.copy()

print("Working Copy Created:", df_clean.shape)
print("Original Preserved:  ", df.shape)

Working Copy Created: (500, 14)
Original Preserved:   (500, 14)


In [1815]:
# TO FIX DATA TYPES

# unit_pric: strip the ₦ symbol, convert to float
print(f"Before - dtype:", df_clean["unit_price"].dtype)
print(f"Before - sample:", df_clean["unit_price"].head().tolist())

# Step 1: Remove the ₦ Symbol
df_clean["unit_price"] = df_clean["unit_price"].str.replace("₦", "", regex=False)

# Step 2: Convert to Numeric (errors='coerce' turns any remaining bad values to NaN)
df_clean["unit_price"] = pd.to_numeric(df_clean["unit_price"], errors='coerce')

print()
print("After - dtype:", df_clean["unit_price"].dtype)
print("After - sample:", df_clean["unit_price"].head().tolist())
print("Any New Missing Values?", df_clean["unit_price"].isnull().sum())


Before - dtype: str
Before - sample: ['₦5358', '₦11689', '₦23668', '₦15892', '₦16038']

After - dtype: int64
After - sample: [5358, 11689, 23668, 15892, 16038]
Any New Missing Values? 0


In [1816]:
# TO FIX MISSING VALUES

# Phone: Can't Recover a Missing Phone Number
# Fill it with a Placeholder - Keep the Row, Flags the Gap
df_clean["phone"] = df_clean["phone"].fillna("Not Provided")

print("Missing Phone Values After Fill:", df_clean["phone"].isnull().sum())
print("Sample Pnone Values:")
print(df_clean["phone"].value_counts().head())

print()
print()

# total_amount: this is our key revenue metric
# We cannot analyse revenue with missing amounts - drop those rows or calculate total_amount again

# before_shape = df_clean.shape
# df_clean = df_clean.dropna(subset=["total_amount"])
# after_shape = df_clean.shape

# print(f"Before Dropna: {before_shape[0]} rows")
# print(f"After Dropna:  {after_shape[0]} rows")
# print(f"Rows Removed:  {before_shape[0] - after_shape[0]}")
# print(f"Missing Total_Amount: {df_clean['total_amount'].isnull().sum()}")

# TO CALCULATE TOTAL_AMOUNT MISSING VALUES INSTEAD OF DROPPING
df_clean["total_amount"] = df_clean["total_amount"].fillna(
    df_clean["quantity"] * df_clean["unit_price"]
)

# THIS IS USED INCASE IF ANY COLUMN IN QUANTITY & UNIT_PRICE IS MISSING. IT SHOULD DROP THE TOTAL AMOUNT MISSING
df_clean = df_clean.dropna(subset=["total_amount"])

df_clean.head(20)

Missing Phone Values After Fill: 0
Sample Pnone Values:
phone
Not Provided    8
8131647525.0    2
8160748217.0    2
7065652341.0    2
8036356512.0    2
Name: count, dtype: int64




,order_id,customer_name,phone,email,state,city,product_category,product_name,quantity,unit_price,total_amount,order_date,delivery_status,payment_method
0,ORD-0001,Rukayat Obi,7086001338.0,rukayatobi@hotmail.com,Abuja,Maitama,Fashion,Handbag,2,5358,10716.0,2024-02-08,Shipped,Cash on Delivery
1,ORD-0002,Emeka Obiora,8066155940.0,emekaobiora@hotmail.com,Oyo,Ogbomosho,Home & Kitchen,Laundry Basket,3,11689,35067.0,2024-02-03,Processing,USSD
2,ORD-0003,Adaeze Abubakar,8131647525.0,adaeze_abubakar27@gmail.com,OGUN,Ota,Fashion,Women's Dress,1,23668,23668.0,2024-02-20,Cancelled,Bank Transfer
3,ORD-0004,Rukayat Obiora,8130564139.0,rukayatobiora@hotmail.com,anambra,Awka,Home & Kitchen,Curtains Set,3,15892,47676.0,2024-01-13,Cancelled,Opay Wallet
4,ORD-0005,Segun Chukwu,Not Provided,segunchukwu21@gmail.com,KANO,Fagge,Fashion,Sunglasses,5,16038,80190.0,2024-06-13,Shipped,Cash on Delivery
5,ORD-0006,Ifeanyi Obiora,8166270482.0,ifeanyiobiora14@yahoo.com,delta,Agbor,Electronics,USB Charger,5,20986,104930.0,2024-02-05,Delivered,Bank Transfer
6,ORD-0007,Blessing Abubakar,8033911718.0,blessing.abubakar17@yahoo.com,Anambra,Ogidi,Food & Drinks,Lipton Tea Box,1,1716,1716.0,2024-04-09,Processing,Cash on Delivery
7,ORD-0008,Chidi Salami,9057133150.0,chidisalami@yahoo.com,enugu,Awgu,Fashion,Sunglasses,4,24009,96036.0,2024-08-01,Delivered,Card
8,ORD-0009,Ibrahim Nwachukwu,9052997376.0,ibrahim.nwachukwu@hotmail.com,Rivers,Port Harcourt,Food & Drinks,Indomie Noodles (12-pack),5,2749,13745.0,2024-04-12,Shipped,Opay Wallet
9,ORD-0010,Adaeze Lawal,9057262473.0,adaeze.lawal@gmail.com,Delta,Warri,Electronics,Power Bank,3,10661,31983.0,2024-04-21,Processing,Card


In [1817]:
print(df_clean["total_amount"].isnull().sum())

0


In [1818]:
# TO REMOVE DUPLICATES

before_shape = df_clean.shape
df_clean = df_clean.drop_duplicates()
after_shape = df_clean.shape

print(f"Before Drop_Duplicates: {before_shape[0]} rows")
print(f"After_Drop_Duplicates:  {after_shape[0]}")
print(f"Duplicated Removed:     {before_shape[0] - after_shape[0]}")

print()
print()

# Confirm Zero Duplicates Remaining
print(f"Remaining Duplicates: {df_clean.duplicated().sum()}")

# Also Check for Duplicate order_ids (same ID, slightly different rows)
print(f"Duplicate order_ids: ", df_clean.duplicated(subset=["order_id"]).sum())

Before Drop_Duplicates: 500 rows
After_Drop_Duplicates:  500
Duplicated Removed:     0


Remaining Duplicates: 0
Duplicate order_ids:  0


In [1819]:
# TO FIX DATA TYPES

# TO  CONVERT DATE STRING TO DATETIME
# order_date: convert strings to Datetime
print("Before - dtype:", df_clean["order_date"].dtype)
print("Before - sample:", df_clean["order_date"].head(4).tolist())

df_clean["order_date"] = pd.to_datetime(df_clean["order_date"])

print()
print("After - dtype:", df_clean["order_date"].dtype)
print("After - sample:", df_clean["order_date"].head(4).tolist())

print()
print()

# TO CREATE NEW COLUMN DATA FROM ORDER_DATE
# Now we can Extract Data Components - Unlocked By the DataTime Conversion
df_clean["order_month"] = df_clean["order_date"].dt.month
df_clean["order_year"]  = df_clean["order_date"].dt.year

print("Orders Per Month:")
print(df_clean["order_month"].value_counts().sort_index())

print()

print("Orders Per Year:")
print(df_clean["order_year"].value_counts().sort_index())
print()       
print(df_clean["order_year"].unique())             # HOW TO USE *UNIQUE METHOD*


Before - dtype: str
Before - sample: ['2024-02-08', '2024-02-03', '2024-02-20', '2024-01-13']

After - dtype: datetime64[us]
After - sample: [Timestamp('2024-02-08 00:00:00'), Timestamp('2024-02-03 00:00:00'), Timestamp('2024-02-20 00:00:00'), Timestamp('2024-01-13 00:00:00')]


Orders Per Month:
order_month
1     37
2     36
3     43
4     40
5     41
6     39
7     50
8     47
9     37
10    46
11    45
12    39
Name: count, dtype: int64

Orders Per Year:
order_year
2024    500
Name: count, dtype: int64

[2024]


In [1820]:
# TO CLEAN STRING COLUMNS

# State: Standardise To Title Case
print("Before - Unique State Values:")
print(sorted(df_clean["state"].unique()))
print(f"Total Unique Value: {df_clean['state'].nunique()}")  # nunique - Gives the number of unique values.

print()
print()

# To Remove the inconsistency in name in the strings name column
df_clean["state"] = df_clean["state"].str.strip().str.title()

print("After - Unique State Values:")
print(sorted(df_clean["state"].unique()))
print(f"Total Unique Value: {df_clean['state'].nunique()}")

Before - Unique State Values:
['ABUJA', 'ANAMBRA', 'Abuja', 'Anambra', 'DELTA', 'Delta', 'ENUGU', 'Enugu', 'KADUNA', 'KANO', 'Kaduna', 'Kano', 'LAGOS', 'Lagos', 'OGUN', 'OYO', 'Ogun', 'Oyo', 'RIVERS', 'Rivers', 'abuja', 'anambra', 'delta', 'enugu', 'kaduna', 'kano', 'lagos', 'ogun', 'oyo', 'rivers']
Total Unique Value: 30


After - Unique State Values:
['Abuja', 'Anambra', 'Delta', 'Enugu', 'Kaduna', 'Kano', 'Lagos', 'Ogun', 'Oyo', 'Rivers']
Total Unique Value: 10


In [1821]:
# Show the impact on groupby - This is WHY Casing Matters

print("==== BEFORE CLEANING (on raw df) ====")
print(df.groupby("state")["total_amount"].sum().sort_values(ascending=False).head(10))

print()
print()

print("==== AFTER CLEANING (on df_clean) ====")
print(df_clean.groupby("state")["total_amount"].sum().sort_values(ascending=True).head(10))

==== BEFORE CLEANING (on raw df) ====
state
Kaduna    1651629.0
Kano      1612610.0
Rivers    1569941.0
Lagos     1261324.0
Abuja     1160007.0
Ogun      1055678.0
Delta     1026168.0
KANO       994843.0
ABUJA      806734.0
rivers     767064.0
Name: total_amount, dtype: float64


==== AFTER CLEANING (on df_clean) ====
state
Enugu      1326717.0
Oyo        1490897.0
Anambra    1713418.0
Ogun       1815964.0
Delta      2042722.0
Lagos      2077693.0
Abuja      2392020.0
Kaduna     2780271.0
Rivers     2886330.0
Kano       3394017.0
Name: total_amount, dtype: float64


In [1822]:
# TO ADD USEFUL COLUMNS, & DROP UNNECESSARY ONES

# Add a revenue tier label - Useful for Segmentation
def assign_tier(amount):
    if amount > 30000:
        return "Premium"
    if amount > 10000:
        return "Mid-range"
    else:
        return "Standard"

df_clean["order_tier"] = df_clean["total_amount"].apply(assign_tier)

print("Order Tier Distribution:")
print(df_clean["order_tier"].value_counts())

Order Tier Distribution:
order_tier
Premium      235
Mid-range    178
Standard      87
Name: count, dtype: int64


In [1823]:
# Add VAT-inclusive amount (7.5% Nigeria VAT Rate)
df_clean["amount_with_VAT"] = (df_clean["total_amount"] * 0.075).round(2)

print("Sample - total_amount vs amount_with_VAT:")
df_clean[["order_id", "total_amount", "amount_with_VAT"]].head(7)

Sample - total_amount vs amount_with_VAT:


,order_id,total_amount,amount_with_VAT
0,ORD-0001,10716.0,803.70
1,ORD-0002,35067.0,2630.02
2,ORD-0003,23668.0,1775.10
3,ORD-0004,47676.0,3575.70
4,ORD-0005,80190.0,6014.25
5,ORD-0006,104930.0,7869.75
6,ORD-0007,1716.0,128.70


In [1824]:
# FULL VALIDATION

print("====== CLEANING VALIDATION REPORT ======")
print()
print(f"ORIGINAL SHAPE:  {df.shape}")
print(f"CLEAN SHAPE:     {df_clean.shape}")
print(f"ROWS REMOVED:    {df.shape[0] - df_clean.shape[0]} (0 Missing)")

====== CLEANING VALIDATION REPORT ======

ORIGINAL SHAPE:  (500, 14)
CLEAN SHAPE:     (500, 18)
ROWS REMOVED:    0 (0 Missing)


In [1825]:
# To Re-Check the Remaining Missing Value Using .isnull().sum()

remaining_missing = df_clean.isnull().sum()
remaining_missing = remaining_missing[remaining_missing > 0]

if len(remaining_missing) == 0:
    print("Missing Values: NONE ✔")
else:
    print("Missing Values Still Present:")
    print(remaining_missing)

Missing Values: NONE ✔


In [1826]:
# To Re-Check for duplicates 

dupe_count = df_clean.duplicated().sum()
print(f"Duplicate Rows: {dupe_count} {'✔' if dupe_count == 0 else '✖ - Still has duplicates!'}")

print()
print()

# To Re-Check if Data types are correct

print("Data types:")
print(df_clean.dtypes)
print()
print("Unit_Price Is Numeric:", pd.api.types.is_numeric_dtype(df_clean["unit_price"]))
print("Order_Date Is DateTime:", pd.api.types.is_datetime64_any_dtype(df_clean["order_date"]))

Duplicate Rows: 0 ✔


Data types:
order_id                       str
customer_name                  str
phone                       object
email                          str
state                          str
city                           str
product_category               str
product_name                   str
quantity                     int64
unit_price                   int64
total_amount               float64
order_date          datetime64[us]
delivery_status                str
payment_method                 str
order_month                  int32
order_year                   int32
order_tier                     str
amount_with_VAT            float64
dtype: object

Unit_Price Is Numeric: True
Order_Date Is DateTime: True


In [1827]:
print("Unique State ValueS After Cleaning:")
print(sorted(df_clean["state"].unique()))
print(f"Count: {df_clean['state'].nunique()} (Expected 10)")

print()
print()

print("VALUE RANGES:")
print(f"    total_amount - min: NGN {df_clean['total_amount'].min():,.0f},"
      f"max: NGN {df_clean['total_amount'].max():,.0f}")
print(f"    unit_price   - min: NGN {df_clean['unit_price'].min():,.0f},"
      f"max: NGN {df_clean['unit_price'].max():,.0f}")
print(f"    quantity     - min: NGN {df_clean['quantity'].min()},"
      f"max: NGN {df_clean['quantity'].max()}")
print(f"    order_date   - from: {df_clean['order_date'].min().date()},"
      f"to: {df_clean['order_date'].max().date()}")


Unique State ValueS After Cleaning:
['Abuja', 'Anambra', 'Delta', 'Enugu', 'Kaduna', 'Kano', 'Lagos', 'Ogun', 'Oyo', 'Rivers']
Count: 10 (Expected 10)


VALUE RANGES:
    total_amount - min: NGN 842,max: NGN 312,000
    unit_price   - min: NGN 842,max: NGN 44,720
    quantity     - min: NGN 1,max: NGN 5
    order_date   - from: 2024-01-01,to: 2024-12-30


In [1830]:
# To Save the Clean Dataset

df_clean.to_csv("Data/orders_clean.csv", index=False, encoding="utf-8-sig")

print(f"Orders_Clean.csv Saved: {df_clean.shape[0]} Rows x {df_clean.shape[1]} Columns")

# Quick Reload To Verify
verify = pd.read_csv("Data/orders_clean.csv", encoding='utf-8-sig')
print(f"Reload Verified:          {verify.shape[0]} Rows x {verify.shape[1]} Columns")
print("Columns:", verify.columns.tolist())

df_clean.head(7)

Orders_Clean.csv Saved: 500 Rows x 18 Columns
Reload Verified:          500 Rows x 18 Columns
Columns: ['order_id', 'customer_name', 'phone', 'email', 'state', 'city', 'product_category', 'product_name', 'quantity', 'unit_price', 'total_amount', 'order_date', 'delivery_status', 'payment_method', 'order_month', 'order_year', 'order_tier', 'amount_with_VAT']


,order_id,customer_name,phone,email,state,city,product_category,product_name,quantity,unit_price,total_amount,order_date,delivery_status,payment_method,order_month,order_year,order_tier,amount_with_VAT
0,ORD-0001,Rukayat Obi,7086001338.0,rukayatobi@hotmail.com,Abuja,Maitama,Fashion,Handbag,2,5358,10716.0,2024-02-08,Shipped,Cash on Delivery,2,2024,Mid-range,803.70
1,ORD-0002,Emeka Obiora,8066155940.0,emekaobiora@hotmail.com,Oyo,Ogbomosho,Home & Kitchen,Laundry Basket,3,11689,35067.0,2024-02-03,Processing,USSD,2,2024,Premium,2630.02
2,ORD-0003,Adaeze Abubakar,8131647525.0,adaeze_abubakar27@gmail.com,Ogun,Ota,Fashion,Women's Dress,1,23668,23668.0,2024-02-20,Cancelled,Bank Transfer,2,2024,Mid-range,1775.10
3,ORD-0004,Rukayat Obiora,8130564139.0,rukayatobiora@hotmail.com,Anambra,Awka,Home & Kitchen,Curtains Set,3,15892,47676.0,2024-01-13,Cancelled,Opay Wallet,1,2024,Premium,3575.70
4,ORD-0005,Segun Chukwu,Not Provided,segunchukwu21@gmail.com,Kano,Fagge,Fashion,Sunglasses,5,16038,80190.0,2024-06-13,Shipped,Cash on Delivery,6,2024,Premium,6014.25
5,ORD-0006,Ifeanyi Obiora,8166270482.0,ifeanyiobiora14@yahoo.com,Delta,Agbor,Electronics,USB Charger,5,20986,104930.0,2024-02-05,Delivered,Bank Transfer,2,2024,Premium,7869.75
6,ORD-0007,Blessing Abubakar,8033911718.0,blessing.abubakar17@yahoo.com,Anambra,Ogidi,Food & Drinks,Lipton Tea Box,1,1716,1716.0,2024-04-09,Processing,Cash on Delivery,4,2024,Standard,128.70


# TRANSFORMING DATA